### ISSUE TITLE
Implement array_agg aggregation function

### ISSUE THREAD

**Issue by martint** [facebook/presto#1975]

**Comment by snarayananqubole** [comment]
I can take this one if no one is actively working on it.

**Comment by martint** [comment]
Go for it! No one is currently working on this.

**Comment by snarayananqubole** [comment]
To confirm, this work item is only about adding a new aggregation function and not generic support for ordered aggregates. 

This will be supported:
```
select array_agg(x) from foo group by y;
```

This will not be supported as part of this workitem:
```
select array_agg(x order by z) from foo group by y;
```

**Comment by martint** [comment]
Correct. Support for ordered aggregates is orthogonal to this change and can be added later.

**Comment by snarayananqubole** [comment]
I've got a version working.

```
presto:default> select n_regionkey, array_agg(n_nationkey) from nation_s3_orc group by n_regionkey;
 n_regionkey |        _col1        
-------------+---------------------
           0 | [0, 5, 14, 15, 16]  
           1 | [1, 2, 3, 17, 24]   
           4 | [4, 10, 11, 13, 20] 
           3 | [6, 7, 19, 22, 23]  
           2 | [8, 9, 12, 18, 21]  
(5 rows)

presto:default> select n_regionkey, array_agg(n_name) from nation_s3_orc group by n_regionkey;
 n_regionkey |                       _col1                        
-------------+----------------------------------------------------
           0 | [ALGERIA, ETHIOPIA, KENYA, MOROCCO, MOZAMBIQUE]    
           1 | [ARGENTINA, BRAZIL, CANADA, PERU, UNITED STATES]   
           4 | [EGYPT, IRAN, IRAQ, JORDAN, SAUDI ARABIA]          
           3 | [FRANCE, GERMANY, ROMANIA, RUSSIA, UNITED KINGDOM] 
           2 | [INDIA, INDONESIA, JAPAN, CHINA, VIETNAM]          
(5 rows)
```

I had a question, though. I use a json array in a slice to represent the state. Adding every element requires deserializing the json into an object array and adding an object and serializing it back. Is there a better way?

**Comment by martint** [comment]
Nice!

You may be able to keep a list of entries in the aggregation state and only turn them into a slice that represents an array when it's time to evaluate the intermediate or final value of the aggregation.

In the long term, we need to be able to support arbitrary objects flowing between operators to avoid unnecessary serialization/deserialization for complex structures.

**Comment by snarayananqubole** [comment]
Opened PR https://github.com/facebook/presto/pull/2238

**Comment by snarayananqubole** [comment]
New PR with much faster implementation here https://github.com/facebook/presto/pull/2246

**Comment by nhuber** [comment]
why does my presto say function not registered?

**Comment by haozhun** [comment]
@nhuber  You need presto version 0.101 or up to use this function. Does this query for you?

    select array_agg(x) from (values 1, 2, 3) t(x);

**Comment by sevenseablue** [comment]
think there is a bug ！

Memory Pool general
Total Memory 1.27TB
CPU Time 4.22h
Rows213M
Data Size 42.70GB
Cumulative Memory
631,428.73GB seconds

create table flight_userinfo_jiguan_tmp_1 as select user_name, array_agg(reg_name1)[1] as reg_name1，followed by forty more same array_agg(col)[1] as new_col, 
from a left join b on (a.xx=b.xx) group by a.xxx
b is distinct by b.xx, therefore, the num of result is no more than a.
table a is about 40g
table b is about 2g

don't know whether use much memory or add a num greater than memory used in fact .

**Comment by cberner** [comment]
Yes, the current implementation has a high fixed cost. It should be fixed in 0.148 (yet to be released)
